File: android/app/src/main/kotlin/com/privateagent/private_agent/model/ScreenNode.kt
package com.privateagent.private_agent.model

data class ScreenNode(
    val index: Int,
    val text: String = "",
    val contentDescription: String = "",
    val className: String = "",
    val viewId: String = "",
    val isClickable: Boolean = false,
    val isEditable: Boolean = false,
    val isScrollable: Boolean = false,
    val isCheckable: Boolean = false,
    val isEnabled: Boolean = true,
    val left: Int = 0,
    val top: Int = 0,
    val right: Int = 0,
    val bottom: Int = 0,
    val centerX: Int = (left + right) / 2,
    val centerY: Int = (top + bottom) / 2
) {
    val bestLabel: String
        get() {
            if (text.isNotBlank()) return text
            if (contentDescription.isNotBlank()) return contentDescription
            if (viewId.isNotBlank()) {
                val parts = viewId.split("/")
                return parts.lastOrNull() ?: viewId
            }
            return ""
        }

    val isInteractive: Boolean
        get() = isClickable || isEditable || isScrollable || isCheckable

    val shortClassName: String
        get() {
            val parts = className.split(".")
            return parts.lastOrNull() ?: className
        }

    fun matches(query: String): Boolean {
        val q = query.trim().lowercase()
        if (q.isEmpty()) return false
        val t = text.lowercase()
        val c = contentDescription.lowercase()
        val v = viewId.lowercase()
        return t == q || c == q || t.contains(q) || c.contains(q) || v.contains(q)
    }

    companion object {
        fun fromMap(map: Map<String, Any?>): ScreenNode {
            val idx = (map["index"] as? Number)?.toInt() ?: 0
            val text = map["text"]?.toString().orEmpty()
            val desc = map["contentDescription"]?.toString().orEmpty()
            val className = map["className"]?.toString().orEmpty()
            val viewId = map["viewId"]?.toString().orEmpty()
            val isClickable = map["isClickable"] as? Boolean ?: false
            val isEditable = map["isEditable"] as? Boolean ?: false
            val isScrollable = map["isScrollable"] as? Boolean ?: false
            val isCheckable = map["isCheckable"] as? Boolean ?: false
            val isEnabled = map["isEnabled"] as? Boolean ?: true
            val left = (map["left"] as? Number)?.toInt() ?: 0
            val top = (map["top"] as? Number)?.toInt() ?: 0
            val right = (map["right"] as? Number)?.toInt() ?: 0
            val bottom = (map["bottom"] as? Number)?.toInt() ?: 0
            val centerX = (map["centerX"] as? Number)?.toInt() ?: ((left + right) / 2)
            val centerY = (map["centerY"] as? Number)?.toInt() ?: ((top + bottom) / 2)

            return ScreenNode(
                index = idx,
                text = text,
                contentDescription = desc,
                className = className,
                viewId = viewId,
                isClickable = isClickable,
                isEditable = isEditable,
                isScrollable = isScrollable,
                isCheckable = isCheckable,
                isEnabled = isEnabled,
                left = left,
                top = top,
                right = right,
                bottom = bottom,
                centerX = centerX,
                centerY = centerY
            )
        }
    }
}

File: lib/models/screen_node.dart
/// A flattened accessibility node coming from the native
/// `AgentAccessibilityService.dumpScreen()` call.
class ScreenNode {
  final int index;
  final String text;
  final String contentDescription;
  final String className;
  final String viewId;
  final bool isClickable;
  final bool isEditable;
  final bool isScrollable;
  final bool isCheckable;
  final bool isEnabled;
  final int left;
  final int top;
  final int right;
  final int bottom;
  final int centerX;
  final int centerY;

  const ScreenNode({
    required this.index,
    required this.text,
    required this.contentDescription,
    required this.className,
    required this.viewId,
    required this.isClickable,
    required this.isEditable,
    required this.isScrollable,
    required this.isCheckable,
    required this.isEnabled,
    required this.left,
    required this.top,
    required this.right,
    required this.bottom,
    required this.centerX,
    required this.centerY,
  });

  factory ScreenNode.fromMap(Map<dynamic, dynamic> map) {
    int asInt(Object? v) => v is int ? v : int.tryParse('$v') ?? 0;
    bool asBool(Object? v) => v == true || v == 'true';
    String asStr(Object? v) => v?.toString() ?? '';
    return ScreenNode(
      index: asInt(map['index']),
      text: asStr(map['text']),
      contentDescription: asStr(map['contentDescription']),
      className: asStr(map['className']),
      viewId: asStr(map['viewId']),
      isClickable: asBool(map['isClickable']),
      isEditable: asBool(map['isEditable']),
      isScrollable: asBool(map['isScrollable']),
      isCheckable: asBool(map['isCheckable']),
      isEnabled: asBool(map['isEnabled']),
      left: asInt(map['left']),
      top: asInt(map['top']),
      right: asInt(map['right']),
      bottom: asInt(map['bottom']),
      centerX: asInt(map['centerX']),
      centerY: asInt(map['centerY']),
    );
  }

  /// Best human-readable label for this node.
  String get label => text.isNotEmpty ? text : contentDescription;

  bool get hasLabel => label.isNotEmpty;

  bool get isInteractive =>
      isClickable || isEditable || isScrollable || isCheckable;

  /// Short class name without the package prefix, e.g. `EditText`.
  String get shortClass {
    if (className.isEmpty) return 'View';
    final dot = className.lastIndexOf('.');
    return dot >= 0 ? className.substring(dot + 1) : className;
  }

  /// True when the node matches [query] by exact label, case-insensitive
  /// contains on text/contentDescription, or view-id tail.
  bool matches(String query) {
    final q = query.trim().toLowerCase();
    if (q.isEmpty) return false;
    if (text.toLowerCase() == q || contentDescription.toLowerCase() == q) {
      return true;
    }
    if (text.toLowerCase().contains(q) ||
        contentDescription.toLowerCase().contains(q)) {
      return true;
    }
    if (viewId.isNotEmpty && viewId.toLowerCase().endsWith(q)) return true;
    return false;
  }

  @override
  String toString() => 'ScreenNode(#$index "$label" $shortClass @$centerX,$centerY)';
}
